File: lint-arwaky/src-python/agent/lifecycle_state_manager.py
"""Agent lifecycle — startup, shutdown, health management."""

from __future__ import annotations

from ..contract import (
    PipelineOutputAggregate,
    AgentLifecycleAggregate,
    ICommandExecutorPort,
    IJobRegistryPort,
    ServiceContainerAggregate,
)
from ..taxonomy import Duration, ResponseData
from typing import Type, Any
import logging
import time
import os
import platform
import sys

logger = logging.getLogger("agent.lifecycle")


class AgentState(AgentLifecycleAggregate):
    """Track agent lifecycle state and aggregate system health."""

    container: ServiceContainerAggregate
    start_time: float | None = None

    def __init__(self, container: ServiceContainerAggregate) -> None:
        super().__init__(container=container)

    @property
    def _INTERFACE_COMPLIANCE(self) -> Type[PipelineOutputAggregate]:
        return PipelineOutputAggregate

    @property
    def uptime(self) -> Duration:
        if self.start_time is None:
            return Duration(value=0.0)
        return Duration(value=float(time.time() - self.start_time))

    def mark_started(self) -> None:
        self.started = True
        self.start_time = time.time()
        self.status = "running"
        logger.info("Agent started (uptime tracking enabled)")

    async def get_health(self) -> ResponseData:
        """AGGREGATOR: Agent presses the buttons of multiple ports to gather health."""
        result: dict[str, Any] = {
            "lifecycle": {
                "status": self.status,
                "uptime_seconds": round(float(self.uptime), 1),
                "started": self.started,
            },
            "system": {
                "os": platform.system(),
                "python": sys.version.split()[0],
                "cwd": os.getcwd(),
            },
            "components": {},
        }

        # Agent uses ports to gather more info
        try:
            self.container.get(ICommandExecutorPort)
            result["components"]["executor"] = "healthy"
        except Exception:
            result["components"]["executor"] = "unavailable"

        try:
            registry = self.container.get(IJobRegistryPort)
            jobs = await registry.list_jobs()
            result["components"]["jobs"] = {
                "total": len(jobs),
                "running": sum(
                    1 for j in jobs.values() if j.get("status") == "running"
                ),
            }
        except Exception:
            result["components"]["jobs"] = "unavailable"

        return ResponseData(value=result)

    def mark_stopped(self) -> None:
        self.status = "stopped"
        logger.info("Agent stopped")

    def mark_degraded(self) -> None:
        self.status = "degraded"


def get_state(container: ServiceContainerAggregate) -> AgentState:
    """Factory to create a container-aware agent state."""
    return AgentState(container=container)

File: lint-arwaky/src-rust/agent/lifecycle_state_manager.rs
// lifecycle_state_manager — Agent lifecycle: startup, shutdown, health management.
use async_trait::async_trait;
use crate::contract::AgentLifecycleAggregate;
use crate::taxonomy::{Duration, ResponseData, StdOutput, StdError, MetadataVO, ExitCode};
use std::collections::HashMap;
use std::time::Instant;

pub struct AgentState {
    start_time: Option<Instant>,
    pub started: bool,
    pub status: String,
}

impl AgentState {
    pub fn new() -> Self {
        Self {
            start_time: None,
            started: false,
            status: "initializing".to_string(),
        }
    }
}

#[async_trait]
impl AgentLifecycleAggregate for AgentState {
    fn uptime(&self) -> Duration {
        match self.start_time {
            Some(start) => Duration::new(start.elapsed().as_secs_f64()),
            None => Duration::new(0.0),
        }
    }

    fn mark_started(&mut self) {
        self.started = true;
        self.start_time = Some(Instant::now());
        self.status = "running".to_string();
    }

    async fn get_health(&self) -> ResponseData {
        let mut metadata_map = HashMap::new();
        metadata_map.insert("lifecycle".to_string(), serde_json::json!({
            "status": self.status,
            "uptime_seconds": self.uptime().value,
            "started": self.started,
        }));
        metadata_map.insert("system".to_string(), serde_json::json!({
            "os": std::env::consts::OS,
            "cwd": std::env::current_dir().map(|p| p.to_string_lossy().to_string()).unwrap_or_default(),
        }));
        ResponseData::new(
            StdOutput::new(""),
            StdError::new(""),
            ExitCode::new(0),
            MetadataVO::new(metadata_map),
        )
    }

    fn mark_stopped(&mut self) {
        self.status = "stopped".to_string();
    }

    fn mark_degraded(&mut self) {
        self.status = "degraded".to_string();
    }
}

pub fn get_state() -> AgentState {
    AgentState::new()
}
